File: scripts/mrp_pipeline/orchestrator.py
import os
import sys
import json
from datetime import datetime
from typing import Optional, Dict, Any, List

# Core configuration and models
from scripts.mrp_pipeline.core.config import DIR_JOURNAL, DIR_RAW, DIR_STRUCTURED, DIR_ATOMIC, PATH_FEEDBACK
from scripts.mrp_pipeline.core.llm_client import LLMClient
from scripts.mrp_pipeline.models.source_doc import SourceDoc
from scripts.mrp_pipeline.models.structured_doc import StructuredDoc, KeywordDefinition
from scripts.mrp_pipeline.models.atomic_node import AtomicNode, CausalWeb
from scripts.mrp_pipeline.models.plan_item import PlanFile, PlanItem, NewNodeAction, MergeNodeAction

class MRPOrchestrator:
    """
    🔥 Trái tim điều phối: Pipeline State Machine
    Điều phối luồng dữ liệu 6 pha: MAP -> REDUCE -> PLAN -> REFINE -> VERIFY -> COMMIT
    Lưu checkpoint bền vững dưới dạng file để chống crash.
    """
    def __init__(self, source_path: str, skip_verify: bool = False):
        self.source_path = source_path
        self.source_slug = os.path.basename(source_path).replace(".md", "")
        self.skip_verify = skip_verify  # Cờ bỏ qua kiểm toán trong chế độ chạy batch chuyển tiếp

        # Thiết lập biến môi trường cục bộ để định tuyến Mock LLM chính xác
        os.environ["CURRENT_MRP_SOURCE_SLUG"] = self.source_slug

        self.llm = LLMClient()
        self.timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        self.checkpoint_path = os.path.join(DIR_JOURNAL, f"mrp_checkpoint_{self.source_slug}.json")
        self.state = {
            "source_slug": self.source_slug,
            "source_path": self.source_path,
            "current_phase": "MAP",
            "timestamp": self.timestamp,
            "mapped_data": None,
            "reduced_data": None,
            "plan_timestamp": None,
            "refined": False,
            "verified": False,
            "committed": False
        }
        self.load_checkpoint()

    def load_checkpoint(self):
        """Khôi phục bối cảnh nếu phát hiện crash hoặc dở dang."""
        if os.path.exists(self.checkpoint_path):
            try:
                with open(self.checkpoint_path, "r", encoding="utf-8") as f:
                    saved_state = json.load(f)
                    self.state.update(saved_state)
                    self.timestamp = self.state["timestamp"]
                    print(f"🔄 Khôi phục checkpoint! Pha hiện tại: {self.state['current_phase']}")
            except Exception as e:
                print(f"⚠️ Không thể đọc checkpoint: {e}. Tạo mới.")

    def save_checkpoint(self):
        """Ghi nhận tiến độ làm việc trực tiếp vào file log để chống crash."""
        try:
            with open(self.checkpoint_path, "w", encoding="utf-8") as f:
                json.dump(self.state, f, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"⚠️ Lỗi lưu checkpoint: {e}")

    def clear_checkpoint(self):
        """Xóa checkpoint sau khi hoàn thành toàn bộ pipeline."""
        if os.path.exists(self.checkpoint_path):
            try:
                os.remove(self.checkpoint_path)
            except Exception as e:
                print(f"⚠️ Lỗi xóa checkpoint: {e}")

    def run(self) -> bool:
        """
        Kích hoạt State Machine chạy tuần tự.
        """
        print(f"\n=======================================================")
        print(f"🚀 KHỞI CHẠY MRP PIPELINE CHO: [{self.source_slug}]")
        print(f"=======================================================")

        try:
            # Phase MAP
            if self.state["current_phase"] == "MAP":
                if not self.run_map():
                    return False
                self.state["current_phase"] = "REDUCE"
                self.save_checkpoint()

            # Phase REDUCE
            if self.state["current_phase"] == "REDUCE":
                if not self.run_reduce():
                    return False
                self.state["current_phase"] = "PLAN"
                self.save_checkpoint()

            # Phase PLAN
            if self.state["current_phase"] == "PLAN":
                if not self.run_plan():
                    return False
                # DỪNG LẠI tại đây, chuyển trạng thái và chờ duyệt
                print(f"\n⏸️ PIPELINE ĐÃ DỪNG ĐỂ CHỜ DUYỆT.")
                print(f"👉 Vui lòng kiểm tra file kế hoạch tại:")
                print(f"   05_journal/mrp_plan_{self.timestamp}.md")
                print(f"👉 Hãy cập nhật trạng thái hoặc gõ lệnh để tiếp tục.")
                return True

            # Phase REFINE
            if self.state["current_phase"] == "REFINE":
                if not self.run_refine():
                    return False
                self.state["current_phase"] = "VERIFY"
                self.save_checkpoint()

            # Phase VERIFY
            if self.state["current_phase"] == "VERIFY":
                if self.skip_verify:
                    print("  ⏭️ Chế độ Batch chuyển tiếp: Tạm thời bỏ qua kiểm toán đồ thị để tránh báo động giả.")
                    self.state["verified"] = True
                else:
                    if not self.run_verify():
                        return False
                self.state["current_phase"] = "COMMIT"
                self.save_checkpoint()

            # Phase COMMIT
            if self.state["current_phase"] == "COMMIT":
                if not self.run_commit():
                    return False
                self.clear_checkpoint()
                print(f"\n🎉 HOÀN THÀNH MRP PIPELINE THÀNH CÔNG CHO [{self.source_slug}]!")
                return True

        except Exception as e:
            print(f"❌ Lỗi thực thi Pipeline: {e}")
            self.save_checkpoint()
            raise e

        return False

    def run_auto_to_end(self) -> bool:
        """Chạy tự động hoàn toàn từ đầu đến cuối không dừng (auto-approve)."""
        print(f"⚡ Đang chạy tự động hoàn toàn (Auto-Approve Mode)...")
        try:
            if self.state["current_phase"] == "MAP":
                if not self.run_map(): return False
                self.state["current_phase"] = "REDUCE"
                self.save_checkpoint()
            if self.state["current_phase"] == "REDUCE":
                if not self.run_reduce(): return False
                self.state["current_phase"] = "PLAN"
                self.save_checkpoint()
            if self.state["current_phase"] == "PLAN":
                if not self.run_plan(): return False
                # Tự động approve kế hoạch
                plan_timestamp = self.timestamp
                plan_filepath = os.path.join(DIR_JOURNAL, f"mrp_plan_{plan_timestamp}.md")
                if os.path.exists(plan_filepath):
                    with open(plan_filepath, "r", encoding="utf-8") as f:
                        content = f.read()
                    content = content.replace("Trạng thái: `pending`", "Trạng thái: `approved`")
                    with open(plan_filepath, "w", encoding="utf-8") as f:
                        f.write(content)
                self.state["current_phase"] = "REFINE"
                self.save_checkpoint()
            if self.state["current_phase"] == "REFINE":
                if not self.run_refine(): return False
                self.state["current_phase"] = "VERIFY"
                self.save_checkpoint()
            if self.state["current_phase"] == "VERIFY":
                if self.skip_verify:
                    print("  ⏭️ Chế độ Batch chuyển tiếp: Tạm thời bỏ qua kiểm toán đồ thị để tránh báo động giả.")
                    self.state["verified"] = True
                else:
                    if not self.run_verify(): return False
                self.state["current_phase"] = "COMMIT"
                self.save_checkpoint()
            if self.state["current_phase"] == "COMMIT":
                if not self.run_commit(): return False
                self.clear_checkpoint()
                return True
        except Exception as e:
            print(f"❌ Lỗi thực thi tự động Pipeline: {e}")
            raise e
        return False

    def run_map(self) -> bool:
        """Pha MAP: Chắt lọc nội dung tài liệu thô thành StructuredDoc mẫu."""
        from scripts.mrp_pipeline.phases.phase_mapper import PhaseMapper
        mapper = PhaseMapper(self)
        return mapper.execute()

    def run_reduce(self) -> bool:
        """Pha REDUCE: So sánh keywords với atomic nodes hiện tại để tránh trùng lặp."""
        from scripts.mrp_pipeline.phases.phase_reducer import PhaseReducer
        reducer = PhaseReducer(self)
        return reducer.execute()

    def run_plan(self) -> bool:
        """Pha PLAN: Tạo file kế hoạch cho con người duyệt."""
        from scripts.mrp_pipeline.phases.phase_planner import PhasePlanner
        planner = PhasePlanner(self)
        return planner.execute()

    def run_refine(self) -> bool:
        """Pha REFINE: Thực thi tạo/trộn các atomic nodes sau khi được duyệt."""
        from scripts.mrp_pipeline.phases.phase_refiner import PhaseRefiner
        refiner = PhaseRefiner(self)
        return refiner.execute()

    def run_verify(self) -> bool:
        """Pha VERIFY: Kiểm toán liên kết và tính toàn vẹn cây tri thức sau thay đổi."""
        from scripts.mrp_pipeline.phases.phase_verifier import PhaseVerifier
        verifier = PhaseVerifier(self)
        return verifier.execute()

    def run_commit(self) -> bool:
        """Pha COMMIT: Cập nhật INDEX, lưu trữ plan, hoàn tất và báo cáo."""
        from scripts.mrp_pipeline.phases.phase_committer import PhaseCommitter
        committer = PhaseCommitter(self)
        return committer.execute()


class MRPBatchOrchestrator:
    """
    📦 BỘ ĐIỀU PHỐI HÀNG ĐỢI TUẦN TỰ (Sequential Batch Orchestrator)
    - Quét toàn bộ file status: to-process
    - Sắp xếp chronological (Cũ nhất trước)
    - Xử lý tuần tự tích lũy để tri thức hội tụ thống nhất, không trùng lặp chéo.
    """
    def __init__(self, directory: str, auto_approve: bool = False):
        self.directory = directory
        self.auto_approve = auto_approve

    def scan_and_sort_files(self) -> List[str]:
        """Quét và sắp xếp file chronological theo mtime."""
        if not os.path.exists(self.directory):
            return []

        files_to_process = []
        for file in os.listdir(self.directory):
            if file.endswith(".md"):
                filepath = os.path.join(self.directory, file)
                # Đọc lướt qua để check frontmatter status: to-process
                try:
                    with open(filepath, "r", encoding="utf-8") as f:
                        content = f.read()
                    if "status: to-process" in content:
                        mtime = os.path.getmtime(filepath)
                        files_to_process.append((mtime, filepath))
                except Exception:
                    pass

        # Sắp xếp theo mtime tăng dần (cũ nhất đứng trước)
        files_to_process.sort(key=lambda x: x[0])
        return [f[1] for f in files_to_process]

    def run(self) -> bool:
        files = self.scan_and_sort_files()
        if not files:
            print("\n🎉 Không có tài liệu thô nào cần xử lý (status: to-process)!")
            return True

        print(f"\n=======================================================")
        print(f"📦 BẮT ĐẦU CHẠY BATCH TUẦN TỰ CHO {len(files)} FILES")
        print(f"   (Sắp xếp theo thứ tự thời gian sửa đổi cũ -> mới)")
        print(f"=======================================================")

        for i, filepath in enumerate(files, 1):
            filename = os.path.basename(filepath)
            print(f"\n[TIẾN TRÌNH {i}/{len(files)}] ───────────────")
            print(f"👉 Đang xử lý: {filename}")

            # Đặt cờ skip_verify=True để tạm thời bỏ qua báo động giả khi chạy batch trung gian
            orchestrator = MRPOrchestrator(filepath, skip_verify=True)

            # Reset checkpoint cũ nếu chạy batch mới để đảm bảo tính tuần tự sạch
            if not self.auto_approve and orchestrator.state["current_phase"] in ("REFINE", "VERIFY", "COMMIT"):
                # Nếu không auto-approve và checkpoint đang ở nửa sau -> tiếp tục (user vừa approve thủ công)
                pass
            else:
                # Nếu chạy từ đầu hoặc auto-approve -> reset checkpoint
                if os.path.exists(orchestrator.checkpoint_path):
                    os.remove(orchestrator.checkpoint_path)
                orchestrator.state["current_phase"] = "MAP"
                orchestrator.save_checkpoint()

            if self.auto_approve:
                # Chạy một mạch từ đầu đến cuối không dừng
                success = orchestrator.run_auto_to_end()
                if not success:
                    print(f"❌ Lỗi: Chạy tự động thất bại tại file: {filename}")
                    return False
            else:
                # Dừng ở PLAN, chờ duyệt thủ công
                success = orchestrator.run()
                if not success:
                    return False

                # Sau khi xuất plan, chúng ta phải dừng Batch
                print(f"\n⏸️ Hàng đợi Batch tạm dừng tại [{filename}].")
                print(f"👉 Vui lòng duyệt kế hoạch trước khi Batch tự động chuyển sang file tiếp theo.")
                break

        # 🔥 CHỈ CHẠY 1 LẦN KIỂM TOÁN TĨNH TOÀN DIỆN CUỐI CÙNG SAU KHI BATCH HOÀN TẤT
        if self.auto_approve:
            print(f"\n=======================================================")
            print("🕵️‍♂️  BATCH VERIFIER: Bắt đầu kiểm toán đồ thị tri thức tối hậu")
            print(f"=======================================================")
            from sync_rules_and_memory import audit_links, audit_tree_integrity
            try:
                audit_links()
                audit_tree_integrity()
                print("  ✅ Đồ thị tri thức đạt trạng thái an toàn tuyệt đối 100%!")
            except Exception as e:
                print(f"  ❌ Phát hiện lỗi kiểm toán cuối cùng: {e}")
                return False

            print(f"\n🎉 HOÀN THÀNH TOÀN BỘ BÀN BATCH TỰ ĐỘNG THÀNH CÔNG!")
        return True

File: src/core/orchestrator.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import { LLMClient } from './llm.ts';
import { DIR_JOURNAL, DIR_RAW, VAULT_ROOT } from './config.ts';
import { PhaseMapper } from '../phases/mapper.ts';
import { PhaseReducer } from '../phases/reducer.ts';
import { PhasePlanner } from '../phases/planner.ts';
import { PhaseRefiner } from '../phases/refiner.ts';
import { PhaseVerifier, auditLinks, auditTreeIntegrity } from '../phases/verifier.ts';
import { PhaseCommitter } from '../phases/committer.ts';

export class MRPOrchestrator {
  sourcePath: string;
  sourceSlug: string;
  skipVerify: boolean;
  vaultRoot: string;
  llm: LLMClient;
  timestamp: string;
  checkpointPath: string;
  state: {
    source_slug: string;
    source_path: string;
    current_phase: string;
    timestamp: string;
    mapped_data: any;
    reduced_data: any;
    plan_timestamp: any;
    refined: boolean;
    verified: boolean;
    committed: boolean;
    plan_item_data?: any;
  };

  constructor(sourcePath: string, skipVerify = false) {
    this.sourcePath = sourcePath;
    this.sourceSlug = path.basename(sourcePath).replace('.md', '');
    this.skipVerify = skipVerify;
    this.vaultRoot = VAULT_ROOT;

    // Thiết lập biến môi trường cục bộ để định tuyến Mock LLM chính xác
    process.env.CURRENT_MRP_SOURCE_SLUG = this.sourceSlug;

    this.llm = new LLMClient();

    // Tạo timestamp dạng YYYYMMDD_HHMMSS
    const now = new Date();
    const pad = (n: number) => String(n).padStart(2, '0');
    this.timestamp = `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}_${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`;

    this.checkpointPath = path.join(DIR_JOURNAL, `mrp_checkpoint_${this.sourceSlug}.json`);

    this.state = {
      source_slug: this.sourceSlug,
      source_path: this.sourcePath,
      current_phase: 'MAP',
      timestamp: this.timestamp,
      mapped_data: null,
      reduced_data: null,
      plan_timestamp: null,
      refined: false,
      verified: false,
      committed: false,
    };

    this.loadCheckpoint();
  }

  loadCheckpoint(): void {
    if (fs.existsSync(this.checkpointPath)) {
      try {
        const saved = JSON.parse(fs.readFileSync(this.checkpointPath, 'utf-8'));
        Object.assign(this.state, saved);
        this.timestamp = this.state.timestamp;
        console.log(`🔄 Khôi phục checkpoint! Pha hiện tại: ${this.state.current_phase}`);
      } catch (e: any) {
        console.log(`⚠️ Không thể đọc checkpoint: ${e.message}. Tạo mới.`);
      }
    }
  }

  saveCheckpoint(): void {
    try {
      fs.writeFileSync(this.checkpointPath, JSON.stringify(this.state, null, 2), 'utf-8');
    } catch (e: any) {
      console.log(`⚠️ Lỗi lưu checkpoint: ${e.message}`);
    }
  }

  clearCheckpoint(): void {
    if (fs.existsSync(this.checkpointPath)) {
      try {
        fs.unlinkSync(this.checkpointPath);
      } catch (e: any) {
        console.log(`⚠️ Lỗi xóa checkpoint: ${e.message}`);
      }
    }
  }

  async run(): Promise<boolean> {
    console.log(`\n=======================================================`);
    console.log(`🚀 KHỞI CHẠY MRP PIPELINE CHO: [${this.sourceSlug}]`);
    console.log(`=======================================================`);

    try {
      // Phase MAP
      if (this.state.current_phase === 'MAP') {
        const success = await this.runMap();
        if (!success) return false;
        this.state.current_phase = 'REDUCE';
        this.saveCheckpoint();
      }

      // Phase REDUCE
      if (this.state.current_phase === 'REDUCE') {
        const success = await this.runReduce();
        if (!success) return false;
        this.state.current_phase = 'PLAN';
        this.saveCheckpoint();
      }

      // Phase PLAN
      if (this.state.current_phase === 'PLAN') {
        const success = await this.runPlan();
        if (!success) return false;
        this.saveCheckpoint();
        // DỪNG LẠI tại đây, chờ duyệt
        console.log(`\n⏸️ PIPELINE ĐÃ DỪNG ĐỂ CHỜ DUYỆT.`);
        console.log(`👉 Vui lòng kiểm tra file kế hoạch tại:`);
        console.log(`   05_journal/mrp_plan_${this.timestamp}.md`);
        console.log(`👉 Hãy cập nhật trạng thái hoặc gõ lệnh để tiếp tục.`);
        return true;
      }

      // Phase REFINE
      if (this.state.current_phase === 'REFINE') {
        const success = await this.runRefine();
        if (!success) return false;
        this.state.current_phase = 'VERIFY';
        this.saveCheckpoint();
      }

      // Phase VERIFY
      if (this.state.current_phase === 'VERIFY') {
        if (this.skipVerify) {
          console.log('  ⏭️ Chế độ Batch chuyển tiếp: Tạm thời bỏ qua kiểm toán đồ thị để tránh báo động giả.');
          this.state.verified = true;
        } else {
          const success = await this.runVerify();
          if (!success) return false;
        }
        this.state.current_phase = 'COMMIT';
        this.saveCheckpoint();
      }

      // Phase COMMIT
      if (this.state.current_phase === 'COMMIT') {
        const success = await this.runCommit();
        if (!success) return false;
        this.clearCheckpoint();
        console.log(`\n🎉 HOÀN THÀNH MRP PIPELINE THÀNH CÔNG CHO [${this.sourceSlug}]!`);
        return true;
      }
    } catch (e: any) {
      console.log(`❌ Lỗi thực thi Pipeline: ${e.message}`);
      this.saveCheckpoint();
      throw e;
    }

    return false;
  }

  async runAutoToEnd(): Promise<boolean> {
    console.log(`⚡ Đang chạy tự động hoàn toàn (Auto-Approve Mode)...`);
    try {
      if (this.state.current_phase === 'MAP') {
        if (!(await this.runMap())) return false;
        this.state.current_phase = 'REDUCE';
        this.saveCheckpoint();
      }
      if (this.state.current_phase === 'REDUCE') {
        if (!(await this.runReduce())) return false;
        this.state.current_phase = 'PLAN';
        this.saveCheckpoint();
      }
      if (this.state.current_phase === 'PLAN') {
        if (!(await this.runPlan())) return false;
        // Tự động approve kế hoạch
        const planTimestamp = this.timestamp;
        const planFilepath = path.join(DIR_JOURNAL, `mrp_plan_${planTimestamp}.md`);
        if (fs.existsSync(planFilepath)) {
          let content = fs.readFileSync(planFilepath, 'utf-8');
          content = content.replace('Trạng thái: `pending`', 'Trạng thái: `approved`');
          fs.writeFileSync(planFilepath, content, 'utf-8');
        }
        this.state.current_phase = 'REFINE';
        this.saveCheckpoint();
      }
      if (this.state.current_phase === 'REFINE') {
        if (!(await this.runRefine())) return false;
        this.state.current_phase = 'VERIFY';
        this.saveCheckpoint();
      }
      if (this.state.current_phase === 'VERIFY') {
        if (this.skipVerify) {
          console.log('  ⏭️ Chế độ Batch chuyển tiếp: Tạm thời bỏ qua kiểm toán đồ thị để tránh báo động giả.');
          this.state.verified = true;
        } else {
          if (!(await this.runVerify())) return false;
        }
        this.state.current_phase = 'COMMIT';
        this.saveCheckpoint();
      }
      if (this.state.current_phase === 'COMMIT') {
        if (!(await this.runCommit())) return false;
        this.clearCheckpoint();
        return true;
      }
    } catch (e: any) {
      console.log(`❌ Lỗi thực thi tự động Pipeline: ${e.message}`);
      throw e;
    }
    return false;
  }

  async runMap(): Promise<boolean> {
    const mapper = new PhaseMapper(this);
    return mapper.execute();
  }

  async runReduce(): Promise<boolean> {
    const reducer = new PhaseReducer(this);
    return reducer.execute();
  }

  async runPlan(): Promise<boolean> {
    const planner = new PhasePlanner(this);
    return planner.execute();
  }

  async runRefine(): Promise<boolean> {
    const refiner = new PhaseRefiner(this);
    return refiner.execute();
  }

  async runVerify(): Promise<boolean> {
    const verifier = new PhaseVerifier(this);
    return verifier.execute();
  }

  async runCommit(): Promise<boolean> {
    const committer = new PhaseCommitter(this);
    return committer.execute();
  }
}

export class MRPBatchOrchestrator {
  directory: string;
  autoApprove: boolean;

  constructor(directory: string, autoApprove = false) {
    this.directory = directory;
    this.autoApprove = autoApprove;
  }

  scanAndSortFiles(): string[] {
    if (!fs.existsSync(this.directory)) {
      return [];
    }

    const filesToProcess: Array<{ mtime: number; filepath: string }> = [];
    const files = fs.readdirSync(this.directory);

    for (const file of files) {
      if (file.endsWith('.md')) {
        const filepath = path.join(this.directory, file);
        try {
          const content = fs.readFileSync(filepath, 'utf-8');
          if (content.includes('status: to-process')) {
            const stat = fs.statSync(filepath);
            filesToProcess.push({ mtime: stat.mtimeMs, filepath });
          }
        } catch (e) {
          // Bỏ qua nếu có lỗi đọc file
        }
      }
    }

    // Sắp xếp theo mtime tăng dần (cũ nhất đứng trước)
    filesToProcess.sort((a, b) => a.mtime - b.mtime);
    return filesToProcess.map(x => x.filepath);
  }

  async run(): Promise<boolean> {
    const files = this.scanAndSortFiles();
    if (files.length === 0) {
      console.log('\n🎉 Không có tài liệu thô nào cần xử lý (status: to-process)!');
      return true;
    }

    console.log(`\n=======================================================`);
    console.log(`📦 BẮT ĐẦU CHẠY BATCH TUẦN TỰ CHO ${files.length} FILES`);
    console.log(`   (Sắp xếp theo thứ tự thời gian sửa đổi cũ -> mới)`);
    console.log(`=======================================================`);

    for (let i = 0; i < files.length; i++) {
      const filepath = files[i];
      const filename = path.basename(filepath);
      console.log(`\n[TIẾN TRÌNH ${i + 1}/${files.length}] ───────────────`);
      console.log(`👉 Đang xử lý: ${filename}`);

      // Đặt cờ skipVerify=true để tạm thời bỏ qua báo động giả khi chạy batch trung gian
      const orchestrator = new MRPOrchestrator(filepath, true);

      // Reset checkpoint cũ nếu chạy batch mới để đảm bảo tính tuần tự sạch
      if (!this.autoApprove && ['REFINE', 'VERIFY', 'COMMIT'].includes(orchestrator.state.current_phase)) {
        // Nếu không auto-approve và checkpoint đang ở nửa sau -> tiếp tục (user vừa approve thủ công)
      } else {
        // Nếu chạy từ đầu hoặc auto-approve -> reset checkpoint
        if (fs.existsSync(orchestrator.checkpointPath)) {
          fs.unlinkSync(orchestrator.checkpointPath);
        }
        orchestrator.state.current_phase = 'MAP';
        orchestrator.saveCheckpoint();
      }

      if (this.autoApprove) {
        // Chạy một mạch từ đầu đến cuối không dừng
        const success = await orchestrator.runAutoToEnd();
        if (!success) {
          console.log(`❌ Lỗi: Chạy tự động thất bại tại file: ${filename}`);
          return false;
        }
      } else {
        // Dừng ở PLAN, chờ duyệt thủ công
        const success = await orchestrator.run();
        if (!success) {
          return false;
        }

        // Sau khi xuất plan, chúng ta phải dừng Batch
        console.log(`\n⏸️ Hàng đợi Batch tạm dừng tại [${filename}].`);
        console.log(`👉 Vui lòng duyệt kế hoạch trước khi Batch tự động chuyển sang file tiếp theo.`);
        break;
      }
    }

    // 🔥 CHỈ CHẠY 1 LẦN KIỂM TOÁN TĨNH TOÀN DIỆN CUỐI CÙNG SAU KHI BATCH HOÀN TẤT
    if (this.autoApprove) {
      console.log(`\n=======================================================`);
      console.log('🕵️‍♂️  BATCH VERIFIER: Bắt đầu kiểm toán đồ thị tri thức tối hậu');
      console.log(`=======================================================`);
      try {
        const { brokenLinks, portabilityViolations } = auditLinks();
        const inconsistencies = auditTreeIntegrity();

        if (brokenLinks === 0 && portabilityViolations === 0 && inconsistencies === 0) {
          console.log('  ✅ Đồ thị tri thức đạt trạng thái an toàn tuyệt đối 100%!');
        } else {
          throw new Error('Đồ thị chứa liên kết hỏng hoặc không nhất quán.');
        }
      } catch (e: any) {
        console.log(`  ❌ Phát hiện lỗi kiểm toán cuối cùng: ${e.message}`);
        return false;
      }

      console.log(`\n🎉 HOÀN THÀNH TOÀN BỘ BÀN BATCH TỰ ĐỘNG THÀNH CÔNG!`);
    }
    return true;
  }
}
